### ISSUE TITLE
Getting Index was outside the bounds of the array error. 

### ISSUE THREAD

**Issue by JayakumarEMIS** [wiltondb/wiltondb#87]
Hi @staticlibs,
       I am getting below issue in my query while trying to execute in babelfish, can you please help me to fix, i tried reindex and all still not working, can you kindly help me to fix the issue

```
SELECT
CASE WHEN P.UnitsPerPack <= 0 THEN 1 ELSE P.UnitsPerPack END AS PackSize
FROM Runtime.Packing.Pack P
INNER JOIN Runtime.Coding.Code C ON P.PackCodeId = C.CodeId
``` 

![image](https://github.com/user-attachments/assets/89fce0c1-79ec-4985-924b-f52c6f97e524)

log:
----
![image](https://github.com/user-attachments/assets/de63b5c3-51ff-4f63-a50d-f55f52103a66)

**Comment by staticlibs** [comment]
@JayakumarEMIS 

Thanks for the log! Something is wrong while sending back the resulting values from this query. I wonder whether you can return the `UnitsPerPack` column casting it to varchar - to inspect whether there are any unusual values there?

**Comment by JayakumarEMIS** [comment]
Hi @staticlibs, seems there is no unusual data in that column because when I use select with top query I can able to getting the result, i have 2600000 records in the table all record seems to be normal and data type of the column is decimal, cn

With top Query:
```
SELECT TOP 1000000
CASE WHEN P.UnitsPerPack <= 0 THEN 1 ELSE P.UnitsPerPack END AS PackSize
FROM Runtime.Packing.Pack P
```
![image](https://github.com/user-attachments/assets/05a4436e-d7a5-4153-a8d2-eed571859342)


Without Top Query:
```
SELECT 
CASE WHEN P.UnitsPerPack <= 0 THEN 1 ELSE P.UnitsPerPack END AS PackSize
FROM Runtime.Packing.Pack P
```
![image](https://github.com/user-attachments/assets/ac2e43c5-9326-4acc-bc57-57e1fef6d092)

**Comment by JayakumarEMIS** [comment]
Hi @staticlibs, any finding for regarding the issue?

**Comment by staticlibs** [comment]
@JayakumarEMIS 

It is in the queue to be looked into, if you can find the exact record in that table that is causing the problem - it can make it easier.

**Comment by JayakumarEMIS** [comment]
Hi @staticlibs, I did few analysis on this issue, kindly check if it is useful for you

the issue column (UnitsPerPack) I have datatype as decimal(10,4) when we query with case when condition, the result column returning decimal(10,8), I think this cause arithmetic overflow exception for certain values, please check

![image](https://github.com/user-attachments/assets/46cf0ed8-6332-4750-b1fe-5cae728cdbc2)

Babelfish:
----------
![image](https://github.com/user-attachments/assets/4abe2e7f-7444-4757-b314-5fe91a511088)


Sql:
-----------
![image](https://github.com/user-attachments/assets/7f8224ac-ffcd-4e29-bba8-93c7b21b5743)

**Comment by staticlibs** [comment]
@JayakumarEMIS 

> I have datatype as decimal(10,4) when we query with case when condition, the result column returning decimal(10,8)

Thanks for the info! Yes, this is useful, will try to reproduce this.

**Comment by JayakumarEMIS** [comment]
Hi @staticlibs, can you able to reproduce the problem

**Comment by staticlibs** [comment]
@JayakumarEMIS 

Yes, reproduced this, it also happens in the latest Babelfish, reported it upstream in [#3137](https://github.com/babelfish-for-postgresql/babelfish_extensions/issues/3137).

The workaround is to add explicit casts for all branches of a `CASE` expression.

**Comment by psrinivasa2** [comment]
Hi @staticlibs ,
Sorry, Couldn't get the workaround, can you please explain with a example workaround?

**Comment by staticlibs** [comment]
@psrinivasa2 

Yes, the problem happens when the incorrect resulting type is resolved for the `CASE` expression. To avoid it you can use explicit casts so both branches have the same type to begin with and no resolve is taking place:

```tsql
CASE WHEN P.UnitsPerPack <= 0 
THEN CAST(1 as decimal(10, 4)) 
ELSE P.UnitsPerPack 
END AS PackSize
```

**Comment by psrinivasa2** [comment]
Thanks @staticlibs
